File: scripts/SetPedidoRetransmitidoOe/setPedidoRetransmitidoOe.js
/**
 * Retransmite pedido de ofício (SetPedidoRetransmitidoOE) no webservice Ofícios da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { envInt, envStr, resolvePath } from "../../lib/onr_env.js";
import {
  hashErrorHint,
  loadLoginConfig,
  loadOficiosSoapConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_oficios.js";
import { validatePedidoForRetransmitidoOE } from "../../lib/onr_oficios_retransmitido.js";

const PREFIX = "OFICIOS_SET_PEDIDO_RETRANSMITIDO_";
const FIELD_ORDER = ["Hash", "IDPedido", "IDCartorio", "Observacoes"];

function resolveIdPedido() {
  const id =
    envInt(`${PREFIX}ID_PEDIDO`) ?? envInt("OFICIOS_ID_PEDIDO");
  if (id === undefined) {
    throw new Error(
      `Defina ${PREFIX}ID_PEDIDO ou OFICIOS_ID_PEDIDO no .env ` +
        "(IDPedido de ListPedidosOE_V2 / GetPedidoOE)."
    );
  }
  return id;
}

function resolveIdCartorio() {
  const id = envInt(`${PREFIX}ID_CARTORIO`);
  if (id === undefined) {
    throw new Error(
      `Defina ${PREFIX}ID_CARTORIO no .env ` +
        "(obtido em ListCartoriosRestransmitirOE)."
    );
  }
  return id;
}

function parseSkipValidarStatus() {
  const raw = envStr(`${PREFIX}SKIP_VALIDAR_STATUS`);
  if (!raw) return false;
  return ["1", "true", "yes", "sim", "s"].includes(String(raw).trim().toLowerCase());
}

function loadConfig() {
  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    idPedido: resolveIdPedido(),
    idCartorio: resolveIdCartorio(),
    observacoes: envStr(`${PREFIX}OBSERVACOES`) ?? "",
    skipValidarStatus: parseSkipValidarStatus(),
    ...loadOficiosSoapConfig(),
  };
}

function buildRequest(cfg, hashValue) {
  const values = {
    Hash: hashValue,
    IDPedido: cfg.idPedido,
    IDCartorio: cfg.idCartorio,
    Observacoes: cfg.observacoes,
  };
  const ordered = Object.fromEntries(
    FIELD_ORDER.filter((key) => key !== "Observacoes" || cfg.observacoes).map(
      (key) => [key, values[key]]
    )
  );
  return ordered;
}

function businessErrorHint(codigo) {
  const hints = {
    12: "IDPedido inválido.",
    13: "IDCartorio inválido — confira ListCartoriosRestransmitirOE (erro 504 se não permitido).",
    51: "Não foi possível obter dados do pedido.",
    52: "Sem permissão para retransmitir este pedido.",
    53: "Pedido já respondido.",
    54: "Apenas Endereço Rua/Edifício/Loteamento (IDTipoPesquisa 1–3).",
    55: "Não foi possível retransmitir o pedido.",
    501: "Não retransmitir para o cartório de origem do pedido.",
    502: "Pedido já respondido.",
    503: "Tipo de pesquisa não permite retransmissão.",
    504: "Cartório informado não é permitido para este pedido.",
  };
  return hints[codigo] ?? null;
}

function extractGetPedidoOeResult(response) {
  return (
    response?.GetPedidoOEResult ??
    response?.GetPedidoOEResponse?.GetPedidoOEResult ??
    response
  );
}

function extractSetPedidoRetransmitidoOeResult(response) {
  return (
    response?.SetPedidoRetransmitidoOEResult ??
    response?.SetPedidoRetransmitidoOEResponse?.SetPedidoRetransmitidoOEResult ??
    response
  );
}

async function createOficiosClient(cfg) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  return soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });
}

function isOnrUnavailableError(err) {
  const status = err?.response?.status;
  const body =
    typeof err?.body === "string"
      ? err.body
      : typeof err?.response?.data === "string"
        ? err.response.data
        : null;
  return status === 503 || (body && !body.trimStart().startsWith("<"));
}

function formatOnrUnavailableError(err) {
  const status = err?.response?.status;
  const body =
    typeof err?.body === "string"
      ? err.body
      : typeof err?.response?.data === "string"
        ? err.response.data
        : null;
  return `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`;
}

async function preflightGetPedidoOe(client, cfg) {
  if (cfg.skipValidarStatus) return 0;

  const baseIndex = envInt("ONR_HASH_TOKEN_INDEX", 0) ?? 0;
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg, {
    tokenIndex: baseIndex,
  });

  let response;
  try {
    [response] = await client.GetPedidoOEAsync({
      oRequest: { Hash: hash, IDPedido: cfg.idPedido },
    });
  } catch (err) {
    if (isOnrUnavailableError(err)) {
      throw new Error(formatOnrUnavailableError(err));
    }
    throw err;
  }

  const pedido = extractGetPedidoOeResult(response);

  if (!pedido?.RETORNO) {
    console.error(
      `GetPedidoOE falhou: [${pedido?.CODIGOERRO}] ${pedido?.ERRODESCRICAO}`
    );
    const hint =
      businessErrorHint(pedido?.CODIGOERRO) ??
      hashErrorHint(pedido?.CODIGOERRO);
    if (hint) console.error(hint);
    return 1;
  }

  const warnings = validatePedidoForRetransmitidoOE(pedido);
  if (!warnings.length) {
    console.log(
      `Pré-validação OK — pedido ${cfg.idPedido}: IDStatus=${pedido.IDStatus ?? "—"}, ` +
        `IDTipoPesquisa=${pedido.IDTipoPesquisa ?? "—"} ` +
        `(protocolo ${pedido.Protocolo ?? "—"}).`
    );
    return 0;
  }

  console.error(
    "=== Pré-validação GetPedidoOE — pedido inelegível para retransmissão ==="
  );
  for (const w of warnings) {
    console.error(`  • ${w}`);
  }
  console.error(
    "\nSetPedidoRetransmitidoOE não será chamado. Use pedido Aberto com IDTipoPesquisa 1–3. " +
      `Cartórios permitidos: ListCartoriosRestransmitirOE. Para ignorar: ${PREFIX}SKIP_VALIDAR_STATUS=true`
  );
  return 1;
}

async function setPedidoRetransmitidoOe(client, oRequest) {
  let response;
  try {
    [response] = await client.SetPedidoRetransmitidoOEAsync({ oRequest });
  } catch (err) {
    if (isOnrUnavailableError(err)) {
      throw new Error(formatOnrUnavailableError(err));
    }
    throw err;
  }

  return extractSetPedidoRetransmitidoOeResult(response);
}

async function main() {
  const cfg = loadConfig();
  const client = await createOficiosClient(cfg);

  const baseIndex = envInt("ONR_HASH_TOKEN_INDEX", 0) ?? 0;
  const preflightRan = !cfg.skipValidarStatus;

  if (preflightRan && (await preflightGetPedidoOe(client, cfg)) !== 0) {
    process.exit(1);
  }

  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg, {
    tokenIndex: preflightRan ? baseIndex + 1 : baseIndex,
  });
  const oRequest = buildRequest(cfg, hash);

  console.log("\n=== Parâmetros SetPedidoRetransmitidoOE ===");
  console.log(JSON.stringify(oRequest, null, 2));
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await setPedidoRetransmitidoOe(client, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nSetPedidoRetransmitidoOE falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    const hint =
      businessErrorHint(response.CODIGOERRO) ?? hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  console.log(
    `\nOK — Pedido ${cfg.idPedido} retransmitido para cartório ${cfg.idCartorio}.`
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: scripts/SetPedidoRetransmitidoOe/setPedidoRetransmitidoOe.py
#!/usr/bin/env python3
"""Retransmite pedido de ofício (SetPedidoRetransmitidoOE) no webservice Ofícios da ONR."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_env import env_int, env_str  # noqa: E402
from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_oficios import (  # noqa: E402
    hash_error_hint,
    load_login_config,
    load_oficios_soap_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_oficios_retransmitido import validate_pedido_for_retransmitido_oe  # noqa: E402
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_zeep_serialize import serialize_result  # noqa: E402

PREFIX = "OFICIOS_SET_PEDIDO_RETRANSMITIDO_"
FIELD_ORDER = ("Hash", "IDPedido", "IDCartorio", "Observacoes")


def _resolve_id_pedido() -> int:
    id_pedido = env_int(f"{PREFIX}ID_PEDIDO")
    if id_pedido is None:
        id_pedido = env_int("OFICIOS_ID_PEDIDO")
    if id_pedido is None:
        raise SystemExit(
            f"Defina {PREFIX}ID_PEDIDO ou OFICIOS_ID_PEDIDO no .env "
            "(IDPedido de ListPedidosOE_V2 / GetPedidoOE)."
        )
    return id_pedido


def _resolve_id_cartorio() -> int:
    id_cartorio = env_int(f"{PREFIX}ID_CARTORIO")
    if id_cartorio is None:
        raise SystemExit(
            f"Defina {PREFIX}ID_CARTORIO no .env "
            "(obtido em ListCartoriosRestransmitirOE)."
        )
    return id_cartorio


def _parse_skip_validar_status() -> bool:
    raw = env_str(f"{PREFIX}SKIP_VALIDAR_STATUS")
    if not raw:
        return False
    return raw.strip().lower() in ("1", "true", "yes", "sim", "s")


def load_config() -> dict:
    load_dotenv(ROOT / ".env")

    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        "id_pedido": _resolve_id_pedido(),
        "id_cartorio": _resolve_id_cartorio(),
        "observacoes": env_str(f"{PREFIX}OBSERVACOES") or "",
        "skip_validar_status": _parse_skip_validar_status(),
        **load_oficios_soap_config(),
    }


def build_request(cfg: dict, hash_value: str) -> dict:
    values = {
        "Hash": hash_value,
        "IDPedido": cfg["id_pedido"],
        "IDCartorio": cfg["id_cartorio"],
        "Observacoes": cfg["observacoes"],
    }
    keys = (
        FIELD_ORDER
        if cfg["observacoes"]
        else tuple(k for k in FIELD_ORDER if k != "Observacoes")
    )
    return {key: values[key] for key in keys}


def _preflight_get_pedido_oe(cfg: dict) -> int:
    if cfg["skip_validar_status"]:
        return 0

    base_index = env_int("ONR_HASH_TOKEN_INDEX", 0) or 0
    hash_value = resolve_auth_hash(
        cfg["chave"], cfg["login_cfg"], token_index=base_index
    )
    o_request = {"Hash": hash_value, "IDPedido": cfg["id_pedido"]}

    result = call_operation_from_cfg(cfg, "GetPedidoOE", o_request)
    pedido = to_json_safe(serialize_result(result))

    if not pedido.get("RETORNO"):
        codigo = pedido.get("CODIGOERRO", "?")
        descricao = pedido.get("ERRODESCRICAO", "")
        print(f"GetPedidoOE falhou: [{codigo}] {descricao}", file=sys.stderr)
        if isinstance(codigo, int):
            hint = _business_error_hint(codigo) or hash_error_hint(codigo)
            if hint:
                print(hint, file=sys.stderr)
        return 1

    warnings = validate_pedido_for_retransmitido_oe(pedido)
    if not warnings:
        print(
            f"Pré-validação OK — pedido {cfg['id_pedido']}: "
            f"IDStatus={pedido.get('IDStatus', '—')}, "
            f"IDTipoPesquisa={pedido.get('IDTipoPesquisa', '—')} "
            f"(protocolo {pedido.get('Protocolo', '—')})."
        )
        return 0

    print(
        "=== Pré-validação GetPedidoOE — pedido inelegível para retransmissão ===",
        file=sys.stderr,
    )
    for w in warnings:
        print(f"  • {w}", file=sys.stderr)
    print(
        "\nSetPedidoRetransmitidoOE não será chamado. Use pedido Aberto com IDTipoPesquisa 1–3. "
        "Cartórios permitidos: ListCartoriosRestransmitirOE. Para ignorar: "
        f"{PREFIX}SKIP_VALIDAR_STATUS=true",
        file=sys.stderr,
    )
    return 1


def _business_error_hint(codigo: int) -> str | None:
    hints = {
        12: "IDPedido inválido.",
        13: "IDCartorio inválido — confira ListCartoriosRestransmitirOE (erro 504 se não permitido).",
        51: "Não foi possível obter dados do pedido.",
        52: "Sem permissão para retransmitir este pedido.",
        53: "Pedido já respondido.",
        54: "Apenas Endereço Rua/Edifício/Loteamento (IDTipoPesquisa 1–3).",
        55: "Não foi possível retransmitir o pedido.",
        501: "Não retransmitir para o cartório de origem do pedido.",
        502: "Pedido já respondido.",
        503: "Tipo de pesquisa não permite retransmissão.",
        504: "Cartório informado não é permitido para este pedido.",
    }
    return hints.get(codigo)


def main() -> int:
    cfg = load_config()

    base_index = env_int("ONR_HASH_TOKEN_INDEX", 0) or 0
    preflight_ran = not cfg["skip_validar_status"]

    if preflight_ran and _preflight_get_pedido_oe(cfg) != 0:
        return 1

    hash_value = resolve_auth_hash(
        cfg["chave"],
        cfg["login_cfg"],
        token_index=base_index + 1 if preflight_ran else base_index,
    )
    o_request = build_request(cfg, hash_value)

    print("=== Parâmetros SetPedidoRetransmitidoOE ===")
    print(json.dumps(o_request, ensure_ascii=False, indent=2))
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "SetPedidoRetransmitidoOE", o_request)
    response = to_json_safe(serialize_result(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(
            f"\nSetPedidoRetransmitidoOE falhou: [{codigo}] {descricao}",
            file=sys.stderr,
        )
        if isinstance(codigo, int):
            hint = _business_error_hint(codigo) or hash_error_hint(codigo)
            if hint:
                print(hint, file=sys.stderr)
        return 1

    print(
        f"\nOK — Pedido {cfg['id_pedido']} retransmitido para cartório "
        f"{cfg['id_cartorio']}."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
